File: Front-End/refactor.py
import os
import re

def css_to_tailwind(css_str):
    classes = []
    styles = [s.strip() for s in css_str.split(';') if s.strip()]
    
    unhandled = []
    
    for style in styles:
        if ':' not in style:
            continue
        key, val = [p.strip() for p in style.split(':', 1)]
        
        if key == 'display' and val == 'flex': classes.append('flex')
        elif key == 'display' and val == 'grid': classes.append('grid')
        elif key == 'display' and val == 'none': classes.append('hidden')
        elif key == 'display' and val == 'block': classes.append('block')
        elif key == 'display' and val == 'inline-block': classes.append('inline-block')
        elif key == 'align-items' and val == 'center': classes.append('items-center')
        elif key == 'align-items' and val == 'flex-start': classes.append('items-start')
        elif key == 'align-items' and val == 'flex-end': classes.append('items-end')
        elif key == 'justify-content' and val == 'center': classes.append('justify-center')
        elif key == 'justify-content' and val == 'space-between': classes.append('justify-between')
        elif key == 'justify-content' and val == 'flex-start': classes.append('justify-start')
        elif key == 'justify-content' and val == 'flex-end': classes.append('justify-end')
        elif key == 'flex-direction' and val == 'column': classes.append('flex-col')
        elif key == 'flex-shrink' and val == '0': classes.append('shrink-0')
        elif key == 'flex' and val == '1': classes.append('flex-1')
        elif key == 'cursor' and val == 'pointer': classes.append('cursor-pointer')
        elif key == 'font-weight' and val == 'bold': classes.append('font-bold')
        elif key == 'font-weight' and val == '700': classes.append('font-bold')
        elif key == 'font-weight' and val == '600': classes.append('font-semibold')
        elif key == 'font-weight' and val == '500': classes.append('font-medium')
        elif key == 'text-align' and val == 'center': classes.append('text-center')
        elif key == 'text-align' and val == 'left': classes.append('text-left')
        elif key == 'text-align' and val == 'right': classes.append('text-right')
        elif key == 'object-fit' and val == 'cover': classes.append('object-cover')
        elif key == 'overflow' and val == 'hidden': classes.append('overflow-hidden')
        elif key == 'overflow-y' and val == 'auto': classes.append('overflow-y-auto')
        elif key == 'position' and val == 'relative': classes.append('relative')
        elif key == 'position' and val == 'absolute': classes.append('absolute')
        elif key == 'position' and val == 'sticky': classes.append('sticky')
        
        elif key == 'gap':
            if val.endswith('px') or val.endswith('rem') or val.endswith('em'):
                classes.append(f'gap-[{val}]')
            else:
                unhandled.append(style)
        elif key == 'width': classes.append(f'w-[{val}]')
        elif key == 'height': classes.append(f'h-[{val}]')
        elif key == 'min-height': classes.append(f'min-h-[{val}]')
        elif key == 'max-width': classes.append(f'max-w-[{val}]')
        elif key == 'margin-top': classes.append(f'mt-[{val}]')
        elif key == 'margin-bottom': classes.append(f'mb-[{val}]')
        elif key == 'margin-left': classes.append(f'ml-[{val}]')
        elif key == 'margin-right': classes.append(f'mr-[{val}]')
        elif key == 'margin' and len(val.split()) == 1: classes.append(f'm-[{val}]')
        elif key == 'margin':
            parts = val.split()
            if len(parts) == 2:
                classes.append(f'my-[{parts[0]}] mx-[{parts[1]}]')
            else:
                unhandled.append(style)
        elif key == 'padding-top': classes.append(f'pt-[{val}]')
        elif key == 'padding-bottom': classes.append(f'pb-[{val}]')
        elif key == 'padding-left': classes.append(f'pl-[{val}]')
        elif key == 'padding-right': classes.append(f'pr-[{val}]')
        elif key == 'padding' and len(val.split()) == 1: classes.append(f'p-[{val}]')
        elif key == 'padding':
            parts = val.split()
            if len(parts) == 2:
                classes.append(f'py-[{parts[0]}] px-[{parts[1]}]')
            else:
                unhandled.append(style)
        elif key == 'font-size': classes.append(f'text-[{val}]')
        elif key == 'color': classes.append(f'text-[{val.replace(" ", "")}]')
        elif key == 'background': classes.append(f'bg-[{val.replace(" ", "")}]')
        elif key == 'background-color': classes.append(f'bg-[{val.replace(" ", "")}]')
        elif key == 'border-radius': classes.append(f'rounded-[{val}]')
        elif key == 'border':
            classes.append(f'[border:{val.replace(" ", "_")}]')
        elif key == 'border-bottom':
            classes.append(f'[border-bottom:{val.replace(" ", "_")}]')
        elif key == 'box-shadow':
            classes.append(f'[box-shadow:{val.replace(" ", "_")}]')
        elif key == 'transition':
            classes.append(f'[transition:{val.replace(" ", "_")}]')
        elif key == 'grid-template-columns':
            classes.append(f'[grid-template-columns:{val.replace(" ", "_")}]')
        else:
            unhandled.append(style)
            
    return classes, unhandled

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    def replacer(match):
        prefix = match.group(1)
        style_content = match.group(2)
        suffix = match.group(3)
        
        tw_classes, unhandled = css_to_tailwind(style_content)
        
        result = prefix
        if tw_classes:
            result += f' __TW_INJECT__="{" ".join(tw_classes)}"'
            
        if unhandled:
            result += f' style="{"; ".join(unhandled)}"'
            
        result += suffix
        return result

    new_content = re.sub(r'(<[^>]+?)\bstyle="([^"]+)"([^>]*>)', replacer, content)
    
    def merge_class(match):
        tag = match.group(0)
        tw_match = re.search(r'__TW_INJECT__="([^"]+)"', tag)
        if not tw_match: return tag
        
        tw_classes = tw_match.group(1)
        tag = tag.replace(tw_match.group(0), '')
        
        class_match = re.search(r'class="([^"]+)"', tag)
        if class_match:
            old_classes = class_match.group(1)
            tag = tag.replace(class_match.group(0), f'class="{old_classes} {tw_classes}"')
        else:
            parts = tag.split(' ', 1)
            if len(parts) == 2:
                tag = f'{parts[0]} class="{tw_classes}" {parts[1]}'
            else:
                tag = f'{tag[:-1]} class="{tw_classes}">'
                
        tag = re.sub(r'\s+', ' ', tag).replace(' >', '>')
        return tag
        
    new_content = re.sub(r'<[^>]+__TW_INJECT__[^>]+>', merge_class, new_content)
    
    if new_content != content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(new_content)
            
    return True

for root, _, files in os.walk('src'):
    for file in files:
        if file.endswith('.html'):
            filepath = os.path.join(root, file)
            process_file(filepath)

File: Front-End/refactor.js
const fs = require('fs');
const path = require('path');

function cssToTailwind(cssStr) {
    const classes = [];
    const styles = cssStr.split(';').map(s => s.trim()).filter(s => s);
    const unhandled = [];
    
    for (const style of styles) {
        if (!style.includes(':')) continue;
        const [key, valRaw] = style.split(/:(.+)/);
        const val = valRaw.trim();
        const k = key.trim();
        
        if (k === 'display' && val === 'flex') classes.push('flex');
        else if (k === 'display' && val === 'grid') classes.push('grid');
        else if (k === 'display' && val === 'none') classes.push('hidden');
        else if (k === 'display' && val === 'block') classes.push('block');
        else if (k === 'display' && val === 'inline-block') classes.push('inline-block');
        else if (k === 'align-items' && val === 'center') classes.push('items-center');
        else if (k === 'align-items' && val === 'flex-start') classes.push('items-start');
        else if (k === 'align-items' && val === 'flex-end') classes.push('items-end');
        else if (k === 'justify-content' && val === 'center') classes.push('justify-center');
        else if (k === 'justify-content' && val === 'space-between') classes.push('justify-between');
        else if (k === 'justify-content' && val === 'flex-start') classes.push('justify-start');
        else if (k === 'justify-content' && val === 'flex-end') classes.push('justify-end');
        else if (k === 'flex-direction' && val === 'column') classes.push('flex-col');
        else if (k === 'flex-shrink' && val === '0') classes.push('shrink-0');
        else if (k === 'flex' && val === '1') classes.push('flex-1');
        else if (k === 'cursor' && val === 'pointer') classes.push('cursor-pointer');
        else if (k === 'font-weight' && val === 'bold') classes.push('font-bold');
        else if (k === 'font-weight' && val === '700') classes.push('font-bold');
        else if (k === 'font-weight' && val === '600') classes.push('font-semibold');
        else if (k === 'font-weight' && val === '500') classes.push('font-medium');
        else if (k === 'text-align' && val === 'center') classes.push('text-center');
        else if (k === 'text-align' && val === 'left') classes.push('text-left');
        else if (k === 'text-align' && val === 'right') classes.push('text-right');
        else if (k === 'object-fit' && val === 'cover') classes.push('object-cover');
        else if (k === 'overflow' && val === 'hidden') classes.push('overflow-hidden');
        else if (k === 'overflow-y' && val === 'auto') classes.push('overflow-y-auto');
        else if (k === 'position' && val === 'relative') classes.push('relative');
        else if (k === 'position' && val === 'absolute') classes.push('absolute');
        else if (k === 'position' && val === 'sticky') classes.push('sticky');
        
        else if (k === 'gap') {
            if (val.endsWith('px') || val.endsWith('rem') || val.endsWith('em')) {
                classes.push(`gap-[${val}]`);
            } else {
                unhandled.push(style);
            }
        }
        else if (k === 'width') classes.push(`w-[${val}]`);
        else if (k === 'height') classes.push(`h-[${val}]`);
        else if (k === 'min-height') classes.push(`min-h-[${val}]`);
        else if (k === 'max-width') classes.push(`max-w-[${val}]`);
        else if (k === 'margin-top') classes.push(`mt-[${val}]`);
        else if (k === 'margin-bottom') classes.push(`mb-[${val}]`);
        else if (k === 'margin-left') classes.push(`ml-[${val}]`);
        else if (k === 'margin-right') classes.push(`mr-[${val}]`);
        else if (k === 'margin' && val.split(' ').length === 1) classes.push(`m-[${val}]`);
        else if (k === 'margin') {
            const parts = val.split(' ');
            if (parts.length === 2) classes.push(`my-[${parts[0]}] mx-[${parts[1]}]`);
            else unhandled.push(style);
        }
        else if (k === 'padding-top') classes.push(`pt-[${val}]`);
        else if (k === 'padding-bottom') classes.push(`pb-[${val}]`);
        else if (k === 'padding-left') classes.push(`pl-[${val}]`);
        else if (k === 'padding-right') classes.push(`pr-[${val}]`);
        else if (k === 'padding' && val.split(' ').length === 1) classes.push(`p-[${val}]`);
        else if (k === 'padding') {
            const parts = val.split(' ');
            if (parts.length === 2) classes.push(`py-[${parts[0]}] px-[${parts[1]}]`);
            else unhandled.push(style);
        }
        else if (k === 'font-size') classes.push(`text-[${val}]`);
        else if (k === 'color') classes.push(`text-[${val.replace(/ /g, '')}]`);
        else if (k === 'background' || k === 'background-color') classes.push(`bg-[${val.replace(/ /g, '_')}]`);
        else if (k === 'border-radius') classes.push(`rounded-[${val}]`);
        else if (k === 'border') classes.push(`[border:${val.replace(/ /g, '_')}]`);
        else if (k === 'border-bottom') classes.push(`[border-bottom:${val.replace(/ /g, '_')}]`);
        else if (k === 'box-shadow') classes.push(`[box-shadow:${val.replace(/ /g, '_')}]`);
        else if (k === 'transition') classes.push(`[transition:${val.replace(/ /g, '_')}]`);
        else if (k === 'grid-template-columns') classes.push(`[grid-template-columns:${val.replace(/ /g, '_')}]`);
        else unhandled.push(style);
    }
    return { classes, unhandled };
}

function processFile(filepath) {
    let content = fs.readFileSync(filepath, 'utf-8');
    let modified = false;

    // Replace style="..."
    const newContent = content.replace(/(<[^>]+?)\bstyle="([^"]+)"([^>]*>)/g, (match, prefix, styleContent, suffix) => {
        const { classes, unhandled } = cssToTailwind(styleContent);
        
        let result = prefix;
        if (classes.length > 0) {
            result += ` __TW_INJECT__="${classes.join(' ')}"`;
        }
        if (unhandled.length > 0) {
            result += ` style="${unhandled.join('; ')}"`;
        }
        result += suffix;
        modified = true;
        return result;
    });

    // Merge __TW_INJECT__ into class="..."
    const finalContent = newContent.replace(/<[^>]+__TW_INJECT__[^>]+>/g, (tag) => {
        const twMatch = tag.match(/__TW_INJECT__="([^"]+)"/);
        if (!twMatch) return tag;
        
        const twClasses = twMatch[1];
        tag = tag.replace(twMatch[0], '');
        
        const classMatch = tag.match(/class="([^"]+)"/);
        if (classMatch) {
            const oldClasses = classMatch[1];
            tag = tag.replace(classMatch[0], `class="${oldClasses} ${twClasses}"`);
        } else {
            const parts = tag.split(' ');
            if (parts.length >= 2) {
                tag = `${parts[0]} class="${twClasses}" ${parts.slice(1).join(' ')}`;
            } else {
                tag = `${tag.slice(0, -1)} class="${twClasses}">`;
            }
        }
        return tag.replace(/\s+/g, ' ').replace(' >', '>');
    });

    if (finalContent !== content) {
        fs.writeFileSync(filepath, finalContent, 'utf-8');
        console.log(`Updated ${filepath}`);
    }
}

function walkDir(dir) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const fullPath = path.join(dir, file);
        if (fs.statSync(fullPath).isDirectory()) {
            walkDir(fullPath);
        } else if (fullPath.endsWith('.html')) {
            processFile(fullPath);
        }
    }
}

walkDir('src');
